sound/soc/sof/intel/mtl.c

```diff
@@ -83,6 +87,10 @@ static bool mtl_dsp_check_sdw_irq(struct snd_sof_dev *sdev)
 
 	/* read Interrupt IP Pointer */
 	hfintipptr = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HFINTIPPTR) & MTL_HFINTIPPTR_PTR_MASK;
+	/* sanity check the irq ptr */
+	if (hfintipptr == MTL_HFINTIPPTR_PTR_MASK)
+		return false;
```

Comment:
this shouldn't happen IMHO @yongzhi1. It's likely that we have a power-up issue somewhere and some parts of the IP are not up when we read that value.

@bardliao @RanderWang please chime in, you're probably the two contributors with the best understanding of the register layout and power dependencies.

Reply:
Thanks @plbossart , will talk offline with @bardliao @RanderWang on the failure scenario.

Reply:
@plbossart we check all types of irq in hda_dsp_interrupt_thread(). Can any interrupt happens before DSP is powered up?
BTW, we have similar check in [hda_common_check_sdw_irq](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda.c#L340) and [hda_dsp_check_ipc_irq](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-ipc.c#L366)

Reply:
It is very likely that we get a interrupt before dsp is powered up. MTL_HFINTIPPTR is also checked by mtl_enable_interrupts but why no error happen ?  Only reason is that mtl_enable_interrupts is not called before the kernel panic but Linux interrupt framework invokes our registered irq function to process interrupt.

I recommend that we check power up state in "hda_dsp_interrupt_thread(int irq, void *context)" and return early to avoid this issue

Reply:
> I recommend that we check power up state (a sw flag) in "hda_dsp_interrupt_thread(int irq, void *context)" and return early to avoid this issue.

can you clarify this suggestion @RanderWang? I am not following the proposal at all.

Reply:
@RanderWang I did some experiements with @yongzhi1 yesterday and it doesnt look like the DSP not being powered up will cause an exception.

Reply:
@ranj063  was mtl_enable_interrupts called ? MTL_HFINTIPPTR is also checked by mtl_enable_interrupts

Reply:
@plbossart  @ranj063 @bardliao  I do a further check that it may be related to our interrupt status check.
we have two registers for irq and only INSTS.GIS is checked.
#define SOF_HDA_INTCTL			0x20 
#define SOF_HDA_INTSTS			0x24

 According  to spec: even if the GIE bit (Global Interrupt Enable in INTCTL) is set to a 0 to disable processor interrupt generation, the GIS bit may still be set to indicate that stream is requesting service.

A recommend check should be like

````c
static irqreturn_t hda_dsp_interrupt_handler(int irq, void *context)
{
	struct snd_sof_dev *sdev = context;

	/*
	 * Get global interrupt status. It includes all hardware interrupt
	 * sources in the Intel HD Audio controller.
	 */
-	if (snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTSTS) &
-	    SOF_HDA_INTSTS_GIS) {
+      intsts = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTSTS);
+      intctl = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTCTL);

+      /* both GIS and GIE are in bit 31, so check GIE & GIS to make sure interrupt is real one or not */
+      intval = (intsts & intctl);
+      if (intval & SOF_HDA_INTSTS_GIS)
````

Reply:
Ack, will test and update.

Reply:
The page fault is caused by the register read when the DSP is off, no?